New Console Logger Test + Discard before Eval

rishab-partha opened the pull request:
# What does this PR do?

New PR for the console logger test, which is now smaller and has documentation, for the discard before eval and not at the end of epoch functionality.

# What issue(s) does this change relate to?

[CO-2152](https://mosaicml.atlassian.net/browse/CO-2152)
[CO-2030](https://mosaicml.atlassian.net/browse/CO-2030
)

[CO-2152]: https://mosaicml.atlassian.net/browse/CO-2152?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[CO-2030]: https://mosaicml.atlassian.net/browse/CO-2030?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

mvpatel2000 commented:
@eracah can you take a look at test?

mvpatel2000 reviewed (commented):
lgtm, will wait for evan's sign off

rishab-partha commented:
Added support for flushing the logger at FIT_END iff the logger was not flushed the same batch (otherwise, the user sees two identical logs, which is very confusing).

Also, updated tests to match.

mvpatel2000 commented on composer/loggers/console_logger.py:
```diff
@@ -96,21 +98,26 @@ def epoch_end(self, state: State, logger: Logger) -> None:
 
     def fit_end(self, state: State, logger: Logger) -> None:
         # Always clear logged metrics so they don't get logged in a subsequent eval call.
+        cur_batch = int(state.timestamp.batch)
+        if self.last_log != cur_batch:
+            self.log_to_console(self.logged_metrics, prefix='Train ',
+                                state=state)  # log at the end of training if you didn't just log
```
can we get rid of `self.last_log` and instead check if `self.logged_metrics` is non-empty, log it?

mvpatel2000 commented on tests/loggers/test_console_logger.py:
```diff
@@ -56,6 +57,8 @@ def test_console_logger_interval(console_logger_test_stream, console_logger_test
     with open(console_logger_test_file_path, 'r') as f:
         lines = f.readlines()
 
+    for line in lines:
+        print(line)
```
```suggestion
```

rishab-partha commented:
Changed the behavior to only print at fit-end only if the last batch did not log. There are longer-term QoL improvements that should be made in the future.

mvpatel2000 commented on composer/loggers/console_logger.py:
```diff
@@ -96,21 +98,26 @@ def epoch_end(self, state: State, logger: Logger) -> None:
-        # discarded.
-        self.logged_metrics = {}
+            self.last_logged_batch = int(state.timestamp.batch)
+            # Clear logged metrics
+            self.logged_metrics = {}
```
```suggestion
            self.logged_metrics = {}. # Clear logged metrics
```
nit

eracah commented on tests/loggers/test_console_logger.py:
```diff
@@ -287,3 +302,57 @@ def test_console_logger_with_a_callback(console_logger_test_stream, console_logg
+
+    num_losses = 1
+    num_metrics_and_losses_per_logging_event = num_metrics + num_losses
+
```
add a comment explaining why it's this formula

eracah commented on tests/loggers/test_console_logger.py:
```diff
@@ -287,3 +302,57 @@ def test_console_logger_with_a_callback(console_logger_test_stream, console_logg
+
+    assert model.train_metrics is not None
+    num_metrics = len(list(model.train_metrics.keys())) if isinstance(model.train_metrics, MetricCollection) else 1
+    num_metrics += len(list(model.parameters())) * 7 + 1  # number from Adam
```
explain that it's 7 metrics per layer

eracah reviewed (approved):
LGTM
